segmentstore/server/src/main/java/io/pravega/segmentstore/server/writer/AggregatedAppendIntegrityChecker.java

```diff
@@ -0,0 +1,180 @@
+     *
+     * @param segmentId Segment related to the Append.
+     * @param offset Append offset.
+     * @param length Append length.
+     * @param hash Hash computed from the original Append contents.
+     */
+    public synchronized void addAppendIntegrityInfo(long segmentId, long offset, long length, long hash) {
+        if (hash == StreamSegmentAppendOperation.NO_HASH) {
+            // No data integrity checks enabled, do nothing.
+            return;
+        }
+        Preconditions.checkArgument(this.segmentId == segmentId, "Appending integrity information for a wrong segment: " + segmentId);
+        this.appendIntegrityInfo.add(new AggregatedAppendIntegrityInfo(offset, length, hash));
+    }
+
+    /**
+     * Validates that the {@link BufferView} contents associated to multiple appends to be flushed to LTS match with the
+     * individual Append hashes that were originally ingested (or recovered). This method also deals with cases in which
+     * an Append may be split across two consecutive flushes.
+     *
+     * @param segmentId Segment related to the Append.
+     * @param offset Initial offset where aggregatedAppends start.
+     * @param aggregatedAppends Data contents of multiple Appends to eb flushed to LTS.
+     * @throws DataCorruptionException If there is a mismatch between the original Append hashes and the contents of the
+     * aggregatedAppends at the same offset.
+     */
+    synchronized void checkAppendIntegrity(long segmentId, long offset, BufferView aggregatedAppends) throws DataCorruptionException {
+        Preconditions.checkArgument(this.segmentId == segmentId, "Appending integrity information for a wrong segment: " + segmentId);
+        if (aggregatedAppends == null || this.appendIntegrityInfo.isEmpty()) {
+            // No data to check, do nothing.
+            return;
+        }
+
+        // Get the appends for queued for this specific Segment.
+        Iterator<AggregatedAppendIntegrityInfo> iterator = this.appendIntegrityInfo.iterator();
+        int accumulatedLength = 0;
+        int checkedBytes = 0;
+        AggregatedAppendIntegrityInfo integrityInfo;
+
+        // If we had a partial append in the previous block, prepend it to the current aggregated appends to check integrity.
+        if (this.previousPartialAppendData.get() != null) {
+            aggregatedAppends = BufferView.builder(2).add(this.previousPartialAppendData.get()).add(aggregatedAppends).build();
+            offset -= this.previousPartialAppendData.get().getLength();
+            this.previousPartialAppendData.set(null);
+        }
+
+        while (iterator.hasNext() && accumulatedLength < aggregatedAppends.getLength()) {
+            integrityInfo = iterator.next();
+            if (integrityInfo.getContentHash() == StreamSegmentAppendOperation.NO_HASH || integrityInfo.getOffset() < offset) {
+                // Old or not hashed operation, just remove.
+                iterator.remove();
```

Comment:
This is the only place I see items being removed from the appendIntegrityInfo. Shouldn't it happen durring the iteration as they are verified.

Reply:
It can happen in both places, actually. But by doing this, I just make sure that everything gets deleted in one step; if not, we may need to delete it on each integrity check but may need some special handling for partial writes that encompass 2 iterations. So I thought that this is the most intuitive way of deleting old integrity info elements.